modules/WordCloudPlugin/src/main/java/net/clementlevallois/wordcloud/TopTermExtractor.java

```diff
@@ -0,0 +1,186 @@
+        2. The result is a list of TextFragment objects
+        3. We keep only the text fragments that are not punctuation signs nor whitespaces
+         */
+        Map<String, List<String>> mapOfNodeIdsToTheirTextFragments = new HashMap();
+
+        StopWordsRemover stopWordsRemoverEN = new StopWordsRemover(3, "en");
+        StopWordsRemover stopWordsRemoverSECONDLANGUAGE = new StopWordsRemover(3, lang);
+
+        Iterator<Map.Entry<String, String>> iteratorOnNodesAndTheirTextualAttribute = textsFromTheAttribute.entrySet().iterator();
+        while (iteratorOnNodesAndTheirTextualAttribute.hasNext()) {
+            try {
+                Map.Entry<String, String> next = iteratorOnNodesAndTheirTextualAttribute.next();
+                String textualAttribute = next.getValue();
+                String nodeId = next.getKey();
+                List<TextFragment> textFragmentsForOneDescription = UmigonTokenizer.tokenize(textualAttribute, languageSpecificLexicon);
+                for (TextFragment oneTextFragment : textFragmentsForOneDescription) {
+                    if (oneTextFragment.getTypeOfTextFragmentEnum() != TypeOfTextFragmentEnum.WHITE_SPACE
+                            & oneTextFragment.getTypeOfTextFragmentEnum() != TypeOfTextFragmentEnum.PUNCTUATION
+                            & !stopWordsRemoverEN.shouldItBeRemoved(oneTextFragment.getOriginalForm())
+                            & !stopWordsRemoverSECONDLANGUAGE.shouldItBeRemoved(oneTextFragment.getOriginalForm())
+                            & oneTextFragment.getOriginalForm().length() > 4) {
+
+                        if (mapOfNodeIdsToTheirTextFragments.containsKey(nodeId)) {
+                            List<String> textFragmentsForThisNode = mapOfNodeIdsToTheirTextFragments.get(nodeId);
+                            textFragmentsForThisNode.add(oneTextFragment.getOriginalForm());
+                            mapOfNodeIdsToTheirTextFragments.put(nodeId, textFragmentsForThisNode);
+                        } else {
+                            List<String> textFragmentsForThisNode = new ArrayList();
+                            textFragmentsForThisNode.add(oneTextFragment.getOriginalForm());
+                            mapOfNodeIdsToTheirTextFragments.put(nodeId, textFragmentsForThisNode);
+                        }
+                    }
+                }
+            } catch (IOException ex) {
+                Exceptions.printStackTrace(ex);
+                return false;
+            }
+        }
+        DataManager.setMapOfNodeIdsToTheirTextFragments(mapOfNodeIdsToTheirTextFragments);
+        return true;
+    }
+
+    public String topTermsExtractorFromSelectedNodes(List<String> nodeIds, int maxNumberOfTerms) {
+        // the multiset will store unique terms from the text we collected and count how many times each term appears
+        Multiset<String> allTextFragmentsFromAllSelectedNodes = new Multiset();
```

Comment:
I think there is a more elegant way to perform this whole method using streams. If I understand correctly, you construct a list of strings, then you sort it and want to retrieve the top list with its number of occurrences.

By prompting chat gpt with "Using Java Stream APIs and taking a list of strings as input, how do I obtain a sorted top-n list with occurrences count?" you should get a good answer.

Reply:
ChatGPT was perfect for that. It removed the need for a Multiset object that I imported from a utils library just for that - so one less dependency for the plugin 💯 .